syntax = "proto3";

package idm;

service IdmService {
  rpc CreateUser(UserBody) returns (Result) {};
  rpc Login(UserData) returns (Result) {};
  rpc ValidateToken(Token) returns (Result) {};
  rpc DestroyToken(Token) returns (Result) {};
  rpc DeleteUser(UserId) returns (Result) {};
  rpc UpdateUser(UserPatch) returns (Result) {};
}

message UserPatch {
  string token = 1;
  int64 id = 2;
  string username = 3;
  string password = 4;
  string action_role = 5;
  string role = 6;
}

message Result {
  int32 status = 1;
  string message = 2;
}

message UserId {
  string token = 1;
  int64 id = 2;
}

message UserData {
  string username = 1;
  string password = 2;
}

message UserBody {
  string token = 1;
  string username = 2;
  string password = 3;
  string role = 4;
}

message Token {
  string token = 1;
}